### PR TITLE
Initial E3D Loader

### PR DESCRIPTION
**Pull request by marcinn**
image
image

Solves #72 

## Includes

* Initial E3DParser
* E3DModel/E3DSubmodel manager + caching
* E3DNodesInstancer
* E3DResourceFormatLoader
* UserSettings dock for Godot Editor
* cParser port (impl. in C++)
* UserSettings singleton
* SConstruct fixes for proper handling addons/ folder
* Added Demo3D and CabinDemo scenes

## Required configuration for loading models

image

## The Idea / Discussion

* This **is not** a complete implementation - E3D parser is not finished and buggy. But this is a good starting point.
* This **is not** a MaSzyna's complete model loader; model loader will require handling MMD files (i.e. models in the MaSzyna can be composed from multiple E3D files)
* Discussion about the architecture: https://github.com/MaSzyna-Reloaded/MaSzyna-API-wrapper/discussions/68
* Discussion about mesh rendering techniques: #30

### CODE REVIEW

File: addons/libmaszyna/e3d/e3d_model_manager.gd [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        for kid in node.get_children():
+            _set_owner_recursive(kid, new_owner)
+
+
+func load_model(data_path:String, filename: String) -> E3DModel:
+    var output:E3DModel
+    var path = UserSettings.get_maszyna_game_dir() + "/" + data_path + "/" + filename + ".e3d"
+
+    # check users cache
+    var cached_path = "user://e3d_res_cache/"+data_path+"__"+filename+".res"
```
Comment (JezSonic) [inline addons/libmaszyna/e3d/e3d_model_manager.gd:19]: We might probably want to store cache for more things so I'd suggest going with
`"user://cache/e3d/res"` or something similar ;)

**Review by JezSonic — COMMENTED** [review]
Overall it looks good. Will do check code etc. soon

**Comment by JezSonic** [comment]
We should discuss some good default material and environment settings to make it look better just by switching some things in the settings. 

I'm not talking about PBR there but maybe enabling SSAO or similar things by default?

**Comment by marcinn** [comment]
> I'm not talking about PBR there but maybe enabling SSAO or similar things by default?

I thought the same. I've tweaked some settings.

File: addons/libmaszyna/e3d/e3d_model_manager.gd (hunk 1 above)
Comment (marcinn) [reply addons/libmaszyna/e3d/e3d_model_manager.gd]: Done

**Review by JezSonic — COMMENTED** [review]
Godot_v4 3-stable_win64_HEduLHpAdM
Shouldn't happen there I guess...

**Review by JezSonic — APPROVED** [review]
Damn this looks cute. Despite the importing issue, awesome work in general., Approved

**Comment by marcinn** [comment]
> Shouldn't happen there I guess...

Yep. I'll fix that later. It is not just about clearing the value. 

Demo_3d.gd is declared as `@tool` for previewing sky changes in the editor, that's why it saves recursively all changed properties. I must create a separate node (initial GameEnvironment or something like that), which will store sky texture name and will set global shader parameter for the sky shader.

**Comment by marcinn** [comment]
Seems to be cleaned now. You can double-check if the scenes are still working properly. 

Sky texture path shouldn't be saved, but other shader parameters will probably saved. This is a general problem with `@tool` scripts and data set in the runtime. Godot Editor is not designed for such cases, so everything it touches should be stored.

Probably the only one solution is by removing `@tool` decoration from custom nodes, and introduce EditorPlugins which will operate on nodes they support. So instead of implementing loading/reloading runtime files in Node scripts, we should do this somewhere else. Nodes should just call factories in _ready(), and EditorPlugins - when edited scene needs this.

If talking about bogies - I will not fix E3D parser for now. I blindly added support for SUB1 chunk, but nothing changed.

**Comment by marcinn** [comment]
Added cabin, fixed e3d parsing (bogies, etc) and transparency issues.

**Review by JezSonic — APPROVED** [review]

**Review by JezSonic — CHANGES_REQUESTED** [review]
Godot_v4 3-stable_win64_cydbuNgq87
Godot_v4 3-stable_win64_KxmazUmW8a
explorer_7l4fg9D6Bm
Code_pMrnHlijvm


Even though some files do exist, godot nor windows explorer can see and read them marking them as if they do not exist. All files though are visible to VSCode and they look okay.

Can not run this for further testing

**Comment by marcinnformula** [comment]
> Can not run this for further testing

please try to compile (scons) and restart godot editor (without saving changes - godot can destroy files).
in case of local changes made by godot, just do git reset --hard

**Comment by marcinn** [comment]
> Can not run this for further testing

please try to compile sources (scons) and restart godot editor
in case of any local changes made by godot editor, just do git reset --hard

**Comment by JezSonic** [comment]
> 
> > Can not run this for further testing
> 
> please try to compile sources (scons) and restart godot editor
> in case of any local changes made by godot editor, just do git reset --hard
> 

The issue still persists after performing a hard reset. Compilation went fine though

**Comment by marcinn** [comment]
> The issue still persists after performing a hard reset. Compilation went fine though

Please ensure that scons copied addons/libmaszyna to demo/addons/libmaszyna

**Review by JezSonic — APPROVED** [review]

**Pull request merged**